src/backend/plaf.rs

```diff
@@ -0,0 +1,507 @@
+        Circuit as cCircuit, Column as cColumn,
+        ColumnType::{Advice as cAdvice, Fixed as cFixed, Halo2Advice, Halo2Fixed},
+        PolyExpr as cPolyExpr,
+    },
+    wit_gen::{GenericTraceContext, TraceWitness},
+};
+
+use num_bigint::BigUint;
+use polyexen::{
+    expr::{get_field_p, Column as pColumn, ColumnKind, ColumnQuery, Expr as pExpr, PlonkVar},
+    plaf::{
+        ColumnFixed, ColumnWitness, Lookup as pLookup, Plaf, Poly as pPoly, Witness as pWitness,
+    },
+};
+
+#[allow(non_snake_case)]
+pub fn chiquito2Plaf<F: PrimeField<Repr = [u8; 32]>, TraceArgs: Clone, StepArgs: Clone>(
+    circuit: cCircuit<F, TraceArgs, StepArgs>,
+    k: u32,
+    debug: bool,
+) -> (Plaf, ChiquitoPlafWitGen<F, TraceArgs, StepArgs>) {
+    let mut chiquito_plaf = ChiquitoPlaf::new(circuit.clone(), debug);
+    let plaf = chiquito_plaf.get_plaf(k);
+    let empty_witness = plaf.gen_empty_witness();
+    let wit_gen = ChiquitoPlafWitGen {
+        trace: circuit.trace,
+        placement: circuit.placement,
+        selector: circuit.selector,
+        step_types: circuit.step_types,
+        plaf_witness: empty_witness,
+        c_column_id_to_p_column_index: chiquito_plaf.c_column_id_to_p_column_index,
+    };
+
+    (plaf, wit_gen)
+}
+
+#[derive(Clone, Debug)]
+pub struct ChiquitoPlaf<F: PrimeField, TraceArgs, StepArgs: Clone> {
+    debug: bool,
+    circuit: cCircuit<F, TraceArgs, StepArgs>,
+    // Chiquito column id doesn't start from zero.
+    // Plaf column index starts from 0 for each column type (advice, fixed, and instance).
+    // Therefore a mapping is needed to convert chiquito column id to plaf index.
+    c_column_id_to_p_column_index: HashMap<u32, usize>,
+    advice_index: usize, // This is Plaf's index of witness (advice) column.
+    fixed_index: usize,  // This is Plaf's index of fixed column.
+}
+
+impl<F: PrimeField<Repr = [u8; 32]>, TraceArgs, StepArgs: Clone>
+    ChiquitoPlaf<F, TraceArgs, StepArgs>
+{
+    // <Repr = [u8; 32]> is required by `from` function in the following line:
+    // `cPolyExpr::Halo2Expr(e) => pExpr::from(e)`.
+    // This function converts a halo2 Expression<F> to a polyexen `Expr<PlonkVar>`.
+    // Therefore, `F: PrimeField<Repr = [u8; 32]>` is required.
+    pub fn new(
+        circuit: cCircuit<F, TraceArgs, StepArgs>,
+        debug: bool,
+    ) -> ChiquitoPlaf<F, TraceArgs, StepArgs> {
+        ChiquitoPlaf {
+            debug,
+            circuit,
+            c_column_id_to_p_column_index: HashMap::new(),
+            advice_index: 0,
+            fixed_index: 0,
+        }
+    }
+
+    pub fn get_plaf(&mut self, k: u32) -> Plaf {
+        let mut plaf = Plaf::default();
+        let p = get_field_p::<F>();
+        plaf.info.p = p;
+
+        plaf.info.num_rows = 2usize.pow(k);
+
+        let mut c_column_id_to_p_column_index = HashMap::<u32, usize>::new();
+        let mut advice_index = 0;
+        let mut fixed_index = 0;
+
+        for column in self.circuit.columns.iter() {
```

Comment:
circuit.columns is a vec and hence each column has an index, you can use that index as the plaf index and then you do not need all the c_column_id_to_p_column_index, advice_index and fixed_index stuff

Reply:
As discussed, this is still needed because Plaf has separate indices advice and fixed (they both start from zero).